hubble/CFP-37472-flowlog-aggregation.md

```diff
@@ -0,0 +1,147 @@
+
+## Non-Goals
+
+* Have pre-determined aggregation levels (ex: low or medium)
+* Aggregate at a cluster level instead of node. For ex: a flow between two pods on two nodes will generate at least one entry on each node. This aggregation option won't aggregate both entries into one.
+
+## Proposal
+
+### Overview
+
+We propose adding a new field in hubble dynamic export configmap called `fieldAggregate`, similar to `fieldMask`. It will be an array of string specifying the fields of [Flow](https://github.com/cilium/cilium/blob/main/api/v1/flow/flow.proto). 
+
+
+### Configmap
+
+```bash
+hubble:
+  export:
+    dynamic:
+      enabled: true
+      config:
+        enabled: true
+        content:
+        - name: "test001"
+          filePath: "/var/run/cilium/hubble/test001.log"
+          fieldMask: []
+          fieldAggregate: []
+          includeFilters: []
+          excludeFilters: []
+          end: "2023-10-09T23:59:59-07:00"
+        - name: "test002"
+          filePath: "/var/run/cilium/hubble/test002.log"
+          fieldMask: ["source.namespace", "source.pod_name", "destination.namespace", "destination.pod_name", "verdict"]
+          fieldAggregate: ["source.pod_name", "destination.pod_name"]
+          includeFilters:
+          - source_pod: ["default/"]
+            event_type:
+            - type: 1
+          - destination_pod: ["frontend/webserver-975996d4c-7hhgt"]
+          excludeFilters: []
+          end: "2023-10-09T23:59:59-07:00"
+        - name: "test003"
+          filePath: "/var/run/cilium/hubble/test003.log"
+          fieldMask: ["source", "destination","verdict"]
+          fieldAggregate: ["source.pod_name", "destination.pod_name", "l4.TCP.destination_port"]
+          includeFilters: []
+          excludeFilters:
+          - destination_pod: ["ingress/"]
+```
+
+The `fieldAggregate` option would allow user to have different level of aggregation across different exporters. This flexibility is necessary as user may want granular logs for sensitive namespaces and more broad signals from others.
+
+## Impacts / Key Questions
+
+### Impact: Log volume
+
+This configuration can cut down log volume by significant amount. This helps store more logs using same disk space on host. Also, less I/O writes on disk.
+
+### Impact: Cost
+
+Reduces ingestion and storage cost of logs, while providing insight into cluster networking. Querying the data is also simplified because of already aggregated data.
+
+### Will this impact existing clusters?
+
+No, this is not a breaking change. By default, the new option won't aggregate any logs. Adding fields to `fieldAggregate` will still write the same `Flow` json structure in the same location - only the number of entries written will differ. Any tooling built to ingest these logs will keep working as expected.
```

Comment:
From the Motivation section:
```
| source_pod | source_port | destination_pod | destination_port | protocol | ingress_flow_count | egress_flow_count |
|------------|-------------|-----------------|------------------|----------|--------------------|-------------------|
| client-1   | 12345       | server          | 80               | TCP      | 4                 | 6                 |
| client-2   | 23456       | server          | 80               | TCP      | 4                 | 6                 |
```
How are the `ingress_flow_count` and `egress_flow_count` computed with the current JSON flow structure?

Reply:
I was thinking of extending Flow using the [extensions](https://github.com/cilium/cilium/blob/81c42ff218655c50c11e5cb36f7b488a42d04f87/api/v1/flow/flow.proto#L129) field.